readthedocs/search/faceted_search.py

```diff
@@ -229,11 +237,72 @@ def query(self, search, query):
         )
 
         all_queries.extend([sections_nested_query, domains_nested_query])
-        final_query = Bool(should=all_queries)
-        search = search.query(final_query)
 
+        final_query = FunctionScore(
+            query=Bool(should=all_queries),
+            script_score=self._get_script_score(),
+        )
+        search = search.query(final_query)
         return search
 
+    def _get_script_score(self):
+        """
+        Gets an ES script to map the page rank to a valid score weight.
+
+        ES expects the rank to be a number greater than 0,
+        but users can set this between [-10, +10].
+        We map that range to [0.01, 2] (21 possible values).
+
+        The first lower rank (0.8) needs to bring the score from the highest boost (sections.title^2)
+        close to the lowest boost (title^1.5), that way exact results take priority:
+
+        - 2.0 * 0.8 = 1.6 (score close to 1.5, but not lower than it)
+        - 1.5 * 0.8 = 1.2 (score lower than 1.5)
+
+        The first higher rank (1.2) needs to bring the score from the lowest boost (title^1.5)
+        close to the highest boost (sections.title^2), that way exact results take priority:
+
+        - 2.0 * 1.3 = 2.6 (score higher thank 2.0)
+        - 1.5 * 1.3 = 1.95 (score close to 2.0, but not higher than it)
+
+        The next lower and higher ranks need to decrease/increase both scores.
+
+        See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-script-score-query.html#field-value-factor  # noqa
+        """
+        ranking = [
+            0.01,
+            0.05,
+            0.1,
+            0.2,
+            0.3,
+            0.4,
+            0.5,
+            0.6,
+            0.7,
+            0.8,
```

Comment:
we can move this 0.8 up if we want to give more priority to the boost (and move the 1.3 down).